README.md
# UniPAGe
A Universal Python Adaptable GUI extension

Copyright © 2017 Ti Leyon

## UniPAGe Tutorial
The Universal Python Adaptable GUI Extensions (UniPAGe) 

UniPAGe is a Python module primary designed to execute scripts under the five major computing platforms (MS Windows, OS X, Android, Linux and IOS) without modification to the source. It uses Kivy as a GUI frame in the first four Operating Systems mentioned earlier and Pythonista under IOS. Testing the demo under different OS will make you realize that the “look and feel” is quasi-identical across all targeted platforms. This is because UniPAGe is not just a bridge between Kivy and Pythonista but a fully adaptable GUI framework designed to feel and look the same under all screen resolutions and orientations.

To accomplish this feat UniPAGe had to take into consideration the fundamental differences between the two programming paradigms. First UniPAGe detects the working environment by “trying” to import a Kivy module. If it succeeds it sets a class level boolean variable named “kivy” to “True”. Otherwise it sets its value to “False” and use it to redirect the creation of GUI elements to specific functions and to activate the given instance. To adapt the GUI components to various screen and window resolutions and orientations the module defines a basic view ratio of 800x600 pixels which incidentally is the default resolution under Kivy. The user must be familiar with absolute positioning of the various constituent of the GUI in order to work with the extension. Every interface must be conceptualized at that level before placing the different parts on the main scenery. Each element must be conceived in the context of the 800 by 600 landscape. The extension will automatically adjust their size and that of their sub-components such as text and glyph according to variations in resolution or orientation. Try changing the original frame by modifying the first tuple parameter of the widget variable under the main section to experiment with the adaptability factor. Currently aspect ratios on most devices including desktops vary within a range of 1.33 to 1.88. Although the extension’s base ratio resides a the bottom of that range, surprisingly, it adapts with no excessive distortion to the full range.That is if you are not using basic drawings in the pictures. However, you can always adjust images ratios within the main framework. Test the demo on different devices and judge for yourself. You may also use a technic that reserves only a portion of the screen as the working area and fill the rest with a monotonic or artistic texture. The usable space is computed from the smaller size in perfect proportion to 800x600 resolution. On phones and tablets I often use that technic by forcing the device into portrait mode and use the smaller size as the horizontal definition while computing the vertical size from the first one base on the same 800x600 proportion. That way I can use the keyboard without covering any part of the working area.

To use UniPAGe you must put the functions that create screen elements (unitext, unibutton, unilabel, unimage, uniframe) into variables. Then create a list of tuples to hold the definitions of the screen elements. The first tuple in that list must contain two values that define the horizontal and vertical sizes of the main window. All the other tuples define various screen components. They contain two main elements: the variable holding the function that creates it and a tuple which contains the parameters for that function. These parameters are distributed in the following order: horizontal position, vertical position, width, height and a character string. That string will serve as the text for the component and in the case of an image it indicates the filename of that image. Buttons may contain a sixth optional parameter that refers to its targeted function. If it is omitted the button will be automatically assigned to a dummy function. At that point create an instance of the uniscreen class which itself is a descendent of unipage. That instance is initialized by using the list of tuples decribed above as its parameter. The window is then initialized by calling the setpage method. Use the showpage method to run the application.

Computers are molds that help create digital universes. Operating systems are the gatekeepers of these universes. Programming languages are the keys to open the gates. Python with its numerous free and open source modules and extensions is definitely a golden key. The five components implemented in this first release are used exclusively in 90% of all programs in all platforms excluding games. Other basic structures such as checkmarks, radio buttons, grids and even scroll views can be easily added to UniPAGe since their implementation in kivy and Pythonista are also quite similar. The main structure for games under Pythonista is the “Scene view”. Kivy also comprises a similar structure called scatter. Someone has also developed an entity-based game engine for Kivy called KivEnt. I hope this modest contribution is of use to the reader.

<a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.



history.md
```UniPAGe``` was first published on April 8 2017 as a demonstration of an adaptable graphic user interface between different platforms. It comprises two Python classes that create identical user interfaces in ```Kivy``` and the ```Pythonista UI``` without changes to the source code. This alongside custom scaling routines effectively allows a single script to run on all 5 major operating system environments (MS Windows, Apple OS X and iOS, Android and Linux) without any modification whatsoever. All updates to ```UniPAGe``` will be published under the name unipage_demo followed by a number corresponding to the chronological order of their creation but not a version number per se. This file will describe the evolution of UniPAGe through the modifications of its core classes in lieu of a formal ChangeLog.

unipage_demo.py

* Original release

unipage_demo1.py

* Promoted xratio and yratio to class level variables
* Added the tuple (0, 0) as a base resolution that will automatically compute an "optimal" viewing window for any device.

unipage_demo3.py

    Major release to address many inconsistancies,
    
    omissions and bugs discovered during intensive
    
    testing on multiple devices acrross all covered
    
    platforms (mainly on iOS and Android).
    
* Set the font size for the ```Kivy``` environment to adapt to the current window resolution
* Created Android windows within the full screen resolution in all orientations because the ```Kivy``` environment responded erratically when the window is set directly.
* Fixed font ratio for the textfields in ```Pythonista UI```
* Fixed the automated mode (0, 0) in the ```Pthonista UI``` for all iOS devices to present an "optimal" sheet (Tested on many iPhone and iPad models).
 * Note: I could not locate a reference for the height of the title bar in a sheet presentation in ```Pythonista```. A factor of 90 times the vertical ratios works on all tested devices although not quite optimal on higher resolution gears.
* Made the "closepage()" function independent of the global "mypage" variable so it could be used along with the UniPAGe classes in an external module if so desired.


unipage_demo.py
# -----------------------------------------------------------------------------
# Name:        UniPaGe
# Purpose:     Adaptable Graphic User Interface for Python
#
# Author:      Ti Leyon
#
# Created:     04/04/2017
# Copyright:   (c) Ti Leyon 2017
# Licence:     K E
# -----------------------------------------------------------------------------


class unipage(object):
    def __init__(self, kivy, screen_size):
        self.kivy = kivy
        self.screen_size = screen_size
        self.unibuttons = []
        self.unitexts = []
        self.unilabels = []
        self.unimages = []
        self.uniframes = []

    def setscreen(self):
        if self.kivy:
            from kivy.uix.floatlayout import FloatLayout
            from kivy.core.window import Window
            self.root = FloatLayout()
            Window.size = self.screen_size
        else:
            import ui
            self.root = ui.View(frame=(0, 0,
                                       self.screen_size[0],
                                       self.screen_size[1]))

    def unibutton(self, params):
        x, y, w, h, text = params[:5]
        xratio = self.screen_size[0] / 800.0
        yratio = self.screen_size[1] / 600.0
        self.unibuttons.append([])
        function = params[5] if len(params) == 6 else nofunction
        if self.kivy:
            from kivy.uix.button import Button
            self.unibuttons[-1] = Button(text=text, size_hint_y=None,
                                         size_hint_x=None,
                                         height=h * yratio,
                                         width=w * xratio,
                                         pos=(x * xratio, y * yratio),
                                         on_press=function)
            self.root.add_widget(self.unibuttons[-1])
        else:
            import ui
            self.unibuttons[-1] = ui.Button(frame=(x * xratio,
                                                   (600 - y - h) * yratio,
                                                   w * xratio, h * yratio),
                                            title=text)
            last_unibutton = self.unibuttons[-1]
            last_unibutton.background_color = (0.4, 0.4, 0.4)
            last_unibutton.action = function
            last_unibutton.height = h * xratio
            last_unibutton.width = w * yratio
            last_unibutton.tint_color = 'white'
            self.root.add_subview(last_unibutton)

    def unitext(self, params):
        x, y, w, h, text = params[:5]
        xratio = self.screen_size[0] / 800.0
        yratio = self.screen_size[1] / 600.0
        self.unitexts.append([])
        if self.kivy:
            from kivy.uix.textinput import TextInput
            self.unitexts[-1] = TextInput(
                id='text' + str(len(self.unitexts) - 1),
                size_hint_y=None,
                size_hint_x=None,
                height=h * yratio,
                width=w * xratio,
                text=text,
                multiline=True,
                pos=(x * xratio, y * yratio))
            self.root.add_widget(self.unitexts[-1])
        else:
            import ui
            self.unitexts[-1] = ui.TextField(
                frame=(x * xratio, (600 - y - h) * yratio,
                       w * xratio, h * yratio))
            last_unitext = self.unitexts[-1]
            last_unitext.bordered = False
            last_unitext.background_color = 'white'
            last_unitext.font = ('<system>', 23 * xratio)
            last_unitext.text = text
            self.root.add_subview(last_unitext)

    def unilabel(self, params):
        x, y, w, h, text = params[:5]
        xratio = self.screen_size[0] / 800.0
        yratio = self.screen_size[1] / 600.0
        self.unilabels.append([])
        if self.kivy:
            from kivy.uix.label import Label
            self.unilabels[-1] = Label(pos=(x * xratio, y * yratio),
                                       size_hint=(1.0, 1.0), halign="left",
                                       valign="bottom", text=text)
            last_unilabel = self.unilabels[-1]
            last_unilabel.bind(size=last_unilabel.setter('text_size'))
            self.root.add_widget(last_unilabel)
        else:
            import ui
            self.unilabels[-1] = ui.Label(frame=(x * xratio,
                                                 (600 - y - h) * yratio,
                                                 w * xratio, h * yratio))
            last_unilabel = self.unilabels[-1]
            last_unilabel.text = text
            last_unilabel.text_color = 'white'
            last_unilabel.alignment = ALIGN_LEFT = True  # TODO: ???
            last_unilabel.font = ('<system>', 18 * xratio)
            self.root.add_subview(last_unilabel)

    def unimage(self, params):
        x, y, w, h, source = params[:5]
        xratio = self.screen_size[0] / 800.0
        yratio = self.screen_size[1] / 600.0
        self.unimages.append([])
        if self.kivy:
            from kivy.uix.image import Image
            self.unimages[-1] = Image(source=source, allow_stretch=True,
                                      size_hint=(None, None),
                                      size=(w * xratio, h * yratio),
                                      pos=(x * xratio, y * yratio))
            self.root.add_widget(self.unimages[-1])
        else:
            import ui
            self.unimages[-1] = ui.ImageView(name='Image',
                                             frame=(x * xratio,
                                                    (600 - y - h) * yratio,
                                                    w * xratio, h * yratio))
            self.root.add_subview(self.unimages[-1])
            self.unimages[-1].image = ui.Image.named(source)

    def uniframe(self, params):
        x, y, w, h, color = params[:5]
        xratio = self.screen_size[0] / 800.0
        yratio = self.screen_size[1] / 600.0
        if self.kivy:
            from kivy.graphics import Color
            from kivy.graphics import Rectangle
            self.root.canvas.add(Color(color[0], color[1], color[2]))
            self.root.canvas.add(Rectangle(pos=(x * xratio, y * yratio),
                                           size=(w * xratio, h * yratio)))
        else:
            import ui
            xratio = self.screen_size[0] / 800.0
            yratio = self.screen_size[1] / 600.0
            self.uniframes.append([])
            self.uniframes[-1] = ui.View(frame=(x * xratio,
                                                (600 - y - h) * yratio,
                                                w * xratio, h * yratio))
            self.uniframes[-1].background_color = (color[0], color[1],
                                                   color[2], 1.0)
            self.root.add_subview(self.uniframes[-1])

    def showpage(self):
        if self.kivy:
            from kivy.base import runTouchApp
            runTouchApp(self.root)
        else:
            self.root.present('sheet')


class uniscreen(unipage):
    screendef = []

    def __init__(self, screendef):
        try:
            from kivy.uix.floatlayout import FloatLayout
            kivy = True
        except:
            import ui
            kivy = False
        unipage.__init__(self, kivy, screendef[0])
        self.setscreen()
        self.screendef = screendef

    def setpage(self):
        for k in range(1, len(self.screendef)):
            self.screendef[k][0](self, self.screendef[k][1])


def closepage(sender):
    if mypage.kivy:
        from kivy.utils import platform as core_platform
        from kivy.core.window import Window
        import sys

        if core_platform == 'android':
            sys.exit()
        else:
            Window.close()
    else:
        mypage.root.close()


def function_1(sender):
    mypage.unitexts[0].text = 'Oh! You clicked my button.'


def nofunction(sender):
    pass


if __name__ == '__main__':
    unilabel = unipage.unilabel
    uniframe = unipage.uniframe
    unitext = unipage.unitext
    unibutton = unipage.unibutton
    unimage = unipage.unimage
    widgets = [(600, 450),
               (uniframe, (0, 0, 600, 450, (.6, .6, .6))),
               (unilabel, (80, 10, 240, 20, 'Hey I am just a simple label.')),
               (unibutton, (40, 40, 100, 40, 'Click me', function_1)),
               (unibutton, (460, 40, 100, 40, 'Close me', closepage)),
               (unitext, (40, 120, 300, 40, 'I am a text field')),
               (unimage, (460, 310, 100, 100, 'insidelogo.png'))]
    mypage = uniscreen(widgets)
    mypage.setpage()
    mypage.showpage()


unipage_demo3.py
#-----------------------------------------------------------------------------
# Name:      UniPaGe
# Purpose:   Adaptable Graphic User Interface for Python
#
# Author:    Ti Leyon
#
# Created:   04/07/2017
# Copyright: (c) Ti Leyon 2017
# Licence:   This work is licensed under the
#            Creative Commons Attribution-ShareAlike 4.0
#            International License.
#            To view a copy of this license, visit:
#            http://creativecommons.org/licenses/by-sa/4.0/
#            or send a letter to:
#            Creative Commons, PO Box 1866, Mountain View, CA 94042, USA.
#-----------------------------------------------------------------------------

class unipage(object):

    def __init__(self, kivy, screen_size):
        self.kivy = kivy
        self.screen_size = screen_size
        self.unibuttons = []
        self.unitexts = []
        self.unilabels = []
        self.unimages = []
        self.uniframes = []
        self.xratio = self.screen_size[0] / 800.0
        self.yratio = self.screen_size[1] / 600.0

    def setscreen(self):
        if self.kivy:
            from kivy.config import Config
            Config.set('input', 'mouse', 'mouse,multitouch_on_demand')
            from kivy.uix.floatlayout import FloatLayout
            from kivy.core.window import Window
            from kivy.utils import platform as core_platform
            self.root = FloatLayout()
            if (self.xratio == 0) or (self.yratio == 0):
                if core_platform == 'android':
                    self.screen_size = Window.size
                    if self.screen_size[0] < self.screen_size[1]:
                        x = self.screen_size[0]
                        y = self.screen_size[0] / 4 * 3
                        self.screen_size = (x, y)
                else:
                    self.screen_size = (800, 600)
                self.xratio = self.screen_size[0] / 800.0
                self.yratio = self.screen_size[1] / 600.0

            if core_platform == 'android':
                Window.softinput_mode = 'pan'
            else:
                Window.size = self.screen_size
        else:
            import ui
            if (self.xratio == 0) or (self.yratio == 0):
                ss1 = ui.get_screen_size()[0]
                ss3 = ui.get_screen_size()[1]
                notoptimal = True
                while notoptimal:
                    if ss1 % 8 == 0:
                        notoptimal = False
                    else:
                        ss1 -= 1
                ss2 = (ss1 / 4) * 3
                title_bar_height = int(ss3 / 600 * 50)
                if ss2 > ss3 - title_bar_height:
                    ss2 = ss3 - title_bar_height
                    notoptimal = True
                    while notoptimal:
                        if ss2 % 6 == 0:
                            notoptimal = False
                        else:
                            ss2 -= 1
                    ss1 = (ss2 / 3) * 4
                self.screen_size = (ss1, ss2)
                self.xratio = ss1 / 800
                self.yratio = ss2 / 600

            self.root = ui.View(frame=(0,0,self.screen_size[0], \
                        self.screen_size[1]))

    def unibutton(self, params):
        self.unibuttons.append([])
        if len(params) == 6:
            function = params[5]
        else:
            function = nofunction
        if self.kivy:
            from kivy.uix.button import Button
            self.unibuttons[len(self.unibuttons) - 1] = Button(
            text = params[4],
            size_hint_y = None,
            size_hint_x = None,
            height = params[3] * self.yratio,
            width = params[2] * self.xratio,
            font_size = 17.5 * self.yratio,
            pos = (params[0] * self.xratio, params[1] * self.yratio),
            on_press = function )
            self.root.add_widget(self.unibuttons[len(self.unibuttons) - 1])
        else:
            import ui
            self.unibuttons[len(self.unibuttons) - 1] = ui.Button(frame= \
                (params[0] * self.xratio, (600 - params[1] - \
                params[3]) * self.yratio, \
                params[2] * self.xratio, params[3] * self.yratio), \
                title = params[4])
            self.unibuttons[len(self.unibuttons) - 1].background_color \
                = (0.4,0.4,0.4)
            self.unibuttons[len(self.unibuttons) - 1].action = function
            self.unibuttons[len(self.unibuttons) - 1].height = params[3] * \
                self.xratio
            self.unibuttons[len(self.unibuttons) - 1].width = params[2] * \
                self.yratio
            self.unibuttons[len(self.unibuttons) - 1].tint_color = 'white'
            self.unibuttons[len(self.unibuttons) - 1].font = ('<system>', \
                17.5 * self.yratio)
            self.root.add_subview(self.unibuttons[len(self.unibuttons) - 1])

    def unitext(self, params):
        self.unitexts.append([])
        if self.kivy:
            from kivy.uix.textinput import TextInput
            self.unitexts[len(self.unitexts) - 1] = TextInput (
            id = 'text' + str(len(self.unitexts) - 1),
            size_hint_y = None,
            size_hint_x = None,
            height = params[3] * self.yratio,
            width = params[2] * self.xratio,
            text = params[4],
            multiline = True,
            font_size = 17.5 * self.yratio,
            pos = (params[0] * self.xratio, params[1] * self.yratio))
            self.root.add_widget(self.unitexts[len(self.unitexts) - 1])
        else:
            import ui
            self.unitexts[len(self.unitexts) - 1] = ui.TextField(frame=
                (params[0] * self.xratio, (600 - params[1] - params[3]) * \
                self.yratio, params[2] * self.xratio, params[3] * self.yratio))
            self.unitexts[len(self.unitexts) - 1].bordered = False
            self.unitexts[len(self.unitexts) - 1].background_color = 'white'
            self.unitexts[len(self.unitexts) - 1].font = ('<system>', 17.5 * \
                self.yratio)
            self.unitexts[len(self.unitexts) - 1].text = params[4]
            self.root.add_subview(self.unitexts[len(self.unitexts) - 1])

    def unilabel(self, params):
        self.unilabels.append([])
        if self.kivy:
            from kivy.uix.label import Label
            self.unilabels[len(self.unilabels) - 1] = Label(pos = \
                (params[0] * self.xratio, params[1] * self.yratio), \
                size_hint=(1.0,1.0), halign="left", \
                valign="bottom", text = params[4])
            self.unilabels[len(self.unilabels) - 1].font_size = 17.5 * \
                self.yratio
            self.unilabels[len(self.unilabels) - 1].bind(size= \
                self.unilabels[len(self.unilabels) - 1].setter('text_size'))
            self.root.add_widget(self.unilabels[len(self.unilabels) - 1])

        else:

            import ui
            self.unilabels[len(self.unilabels) - 1] = ui.Label(frame= \
                (params[0] * self.xratio,  (600 - params[1] - params[3]) * \
                self.yratio, params[2] * self.xratio, params[3] * self.yratio))
            self.unilabels[len(self.unilabels) - 1].text = params[4]
            self.unilabels[len(self.unilabels) - 1].text_color = 'white'
            self.unilabels[len(self.unilabels) - 1].alignment = \
                ui.ALIGN_LEFT
            self.unilabels[len(self.unilabels) - 1].font = ('<system>', 18 * \
                self.yratio)
            self.root.add_subview(self.unilabels[len(self.unilabels) - 1])

    def unimage(self, params):
        self.unimages.append([])
        if self.kivy:
            from kivy.uix.image import Image
            self.unimages[len(self.unimages) - 1] = Image( source= params[4],
                allow_stretch = True, size_hint = (None, None),
                size=(params[2] * self.xratio, params[3] * self.yratio),
                pos=(params[0] * self.xratio, params[1] * self.yratio))

            self.root.add_widget(self.unimages[len(self.unimages) - 1])

        else:
            import ui
            self.unimages[len(self.unimages) - 1] = (ui.ImageView
            (name = 'Image', frame = (params[0] * self.xratio, \
            (600 - params[1] - params[3]) * self.yratio, \
            params[2] * self.xratio, params[3] * self.yratio)))

            self.root.add_subview (self.unimages[len(self.unimages) - 1])
            self.unimages[len(self.unimages) - 1].image = \
                ui.Image.named(params[4])

    def uniframe(self, params):
        if self.kivy:
            from kivy.graphics import Color
            from kivy.graphics import Rectangle
            self.root.canvas.add(Color (params[4][0],params[4][1], \
                params[4][2]))
            self.root.canvas.add(Rectangle(pos = (params[0] * self.xratio, \
                params[1] * self.yratio), size = (params[2] * self.xratio, \
                params[3] * self.yratio)))
        else:
            import ui
            self.uniframes.append([])
            self.uniframes[len(self.uniframes) - 1] = \
                ui.View(frame=(params[0] * self.xratio, \
                (600 - params[1] - params[3]) * self.yratio, \
                params[2] * self.xratio, params[3] * self.yratio))
            self.uniframes[len(self.uniframes) - 1].background_color = \
                (params[4][0],params[4][1], params[4][2],1.0)
            self.root.add_subview(self.uniframes[len(self.uniframes) - 1])

    def showpage(self):
        if self.kivy:
            from kivy.base import runTouchApp
            runTouchApp(self.root)
        else:
            self.root.present('sheet')

class uniscreen(unipage):
    def __init__(self, screendef):
        try:
            from kivy.uix.floatlayout import FloatLayout
            kivy = True
        except:
            import ui
            kivy = False
        unipage.__init__(self, kivy, screendef[0])
        self.setscreen()
        self.screendef = screendef

    def setpage(self):
        for k in range(1, len(self.screendef)):
            self.screendef[k][0](self, self.screendef[k][1])

def closepage(sender):
    try:

        from kivy.utils import platform as core_platform
        from kivy.core.window import Window
        import sys

        if core_platform == 'android':
            sys.exit()
        else:
            Window.close()

    except:

        sender.superview.close()

def nofunction(sender):
    pass

def function_1(sender):
    mypage.unitexts[0].text = 'Oh! You clicked my button.'

if __name__ == '__main__':
    unilabel = unipage.unilabel
    uniframe = unipage.uniframe
    unitext = unipage.unitext
    unibutton = unipage.unibutton
    unimage = unipage.unimage
    widgets = [(0, 0),
        (uniframe,(0, 0, 600, 450,(.6,.6,.6))),
        (unilabel,(40, 300, 240, 20, 'Hey I am just a simple label.')),
        (unibutton,(40, 40, 100, 40, 'Click me', function_1)),
        (unibutton,(460, 40, 100, 40, 'Close me', closepage)),
        (unitext,(40, 120, 300, 40, 'I am a text field')),
        (unimage,(460, 310, 100, 100,'insidelogo.png'))
        ]
    mypage = uniscreen(widgets)
    mypage.setpage()
    if mypage.kivy:
        mypage.unimages[len(mypage.unimages) - 1].source = \
            'data/logo/kivy-icon-512.png'
    else:
        mypage.unimages[len(mypage.unimages) - 1].source = \
            'test:Pythonista'
    mypage.showpage()
